BGP: OSPF: routing extensions to eos_cli_config_ren role

c-po opened the pull request:
# BGP
* peer-groups can be shutdown for maintenance
* peer-groups can have a description
* neighbors can be configured with next-hop-self for iBGP
* prefixes to announce can be specified
* peer-groups and neighbors support weight attribute
* peer-groups and neighbors support configurable timer values
* neighbor update-source can be set individually (required for iBGP neighborship using loopback addresses learned from OSPF)
* bgp: ipv4: add prefix-list support

# OSPF:
* support for originating a default route
* support redistributing static routes into OSPF process

# LLDP:
* add initial LLDP support to eos_cli_config_gen role

Commit:
lldp: add support to eos_cli_config_gen role

The following YAML representation should render the switch config:

  ### LLDP ###
  lldp:
    timer: 120
    holdtime: 240
    management_address: Management1
    vrf: MGMT

  !
  lldp timer 120
  lldp hold-time 240
  lldp management-address Management1
  lldp management-address vrf MGMT
  lldp run
  !

This implements GitHub PullRequest aristanetworks/ansible-avd#155

Commit:
ospf: support for originating a default route

When default_information_originate is set to always a default route will always
be originated from the OSPF process. It will add the following line to the EOS
configuration:

  default-information originate always

YAML example:

  router_ospf:
    process_ids:
      100:
        default_information_originate:
          always: true

The data-model is thus future extensible with e.g. originating a metric,
route-map ...

Commit:
ospf: support redistributing static routes into OSPF process

Static routes can be either redistributed "all at once" or in conjunction with
using a route-map.

* Redistribute all static routes
  redistribute:
    static:

* Redistribute static routes with a route-map
  redistribute:
    static:
      route_map: foo

This YAML design leaves headroom for adding redistribution of additional
protocols.

Commit:
bgp: peer-group: add shutdown support

Support for shutting down an entire BGP peer-group was added.

Commit:
bgp: peer-group: add support to define description

Commit:
bgp: neighbor: support configuration of next-hop-self

Commit:
bgp: ipv4: add support for defining prefixes to announce

Commit:
bgp: peer-group: neighbor: support configuration of weight attribute

Commit:
bgp: peer-group: neighbor: support configuration of timers

aphillipps commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2:
```diff
@@ -40,6 +40,9 @@ router bgp {{ router_bgp.as }}
 {%             if router_bgp.peer_groups[peer_group].maximum_routes is defined and router_bgp.peer_groups[peer_group].maximum_routes is not none %}
    neighbor {{ peer_group }} maximum-routes {{ router_bgp.peer_groups[peer_group].maximum_routes }}
 {%             endif %}
+{%             if router_bgp.peer_groups[peer_group].shutdown is defined and router_bgp.peer_groups[peer_group].shutdown is not none %}
```
Please update this test as follows:
{%             if router_bgp.peer_groups[peer_group].shutdown is defined and router_bgp.peer_groups[peer_group].shutdown == true %}

c-po replied:
Changed accordingly.

Commit:
lldp: support explicit configuration of run statement

If LLDP is disabled (run: false) no other parameter is rendered into the
configuration and the service is explicitly stopped. LLDP is enabled by default
on Arista switches.

Commit:
bgp: neighbor: support configuration of update-source

Commit:
bgp: ipv4: add prefix-list support

A BGP per neighbor prefix-list for in and out can be defined in the
address-family ipv4 block. See README for command definition.

Commit:
bgp: ipv4: support default-originate for individual peers

Send an IPv4 default route downwards the BGP peer.

c-po commented:
Anything todo on my side which is missing?

carlbuchmann commented:
@c-po - not at this time, @aphillipps plans to review the 1st week of August.

Commit:
ospf: support redistributing connected routes

aphillipps commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2:
```diff
@@ -56,6 +56,9 @@ router bgp {{ router_bgp.as }}
 {%             if router_bgp.neighbors[neighbor].remote_as is defined and router_bgp.neighbors[neighbor].remote_as is not none %}
    neighbor {{ neighbor }} remote-as {{ router_bgp.neighbors[neighbor].remote_as }}
 {%             endif %}
+{%             if router_bgp.neighbors[neighbor].next_hop_self is defined and router_bgp.neighbors[neighbor].next_hop_self is not none %}
```
Please change to : {%             if router_bgp.neighbors[neighbor].next_hop_self is defined and router_bgp.neighbors[neighbor].next_hop_self == true %}

c-po replied:
Fixed

aphillipps commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-ospf.j2:
```diff
@@ -25,6 +25,9 @@ router ospf {{ process_id }}
 {%         if router_ospf.process_ids[process_id].max_lsa is defined %}
    max-lsa {{ router_ospf.process_ids[process_id].max_lsa }}
 {%         endif %}
+{%         if router_ospf.process_ids[process_id].default_information_originate is defined and router_ospf.process_ids[process_id].default_information_originate == true %}
```
The name for the key 'default_information_originate' is misleading as the configuration generated has the option 'always'. This section should be refactored with the possibility to add more options in the future (like metrics, metric-types and route-maps) and have 'always' as an option as well.

c-po replied:
Fixed

aphillipps commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-bgp.j2:
```diff
@@ -201,6 +201,11 @@ router bgp {{ router_bgp.as }}
 {%             if router_bgp.address_family_ipv4.neighbors[neighbor].prefix_list_out is defined and router_bgp.address_family_ipv4.neighbors[neighbor].prefix_list_out is not none %}
       neighbor {{ neighbor }} prefix-list {{ router_bgp.address_family_ipv4.neighbors[neighbor].prefix_list_out }} out
 {%             endif %}
+{%             if router_bgp.address_family_ipv4.neighbors[neighbor].default_originate is defined and router_bgp.address_family_ipv4.neighbors[neighbor].default_originate == "always" %}
```
This templating model doesn't cover default originate always with a route-map and does not work. Here is one suggestion for the data-model for which you would then need to change the jinja templating accordingly:
        default_originate:
          always: < true | false >
          route_map: < route_map_name >
Let me know if you have another suggestion for it .

c-po replied:
I can adopt to this data-model

c-po commented:
Hi @aphillipps, I have invorporated all requested changes into this PR and force pushed it.

c-po commented:
Rebased to latest `devel` HEAD

c-po commented:
I was not aware that someone can add commits on another persons PR? How is this possible? I will squash your changes into the original commits incl. the data model change.

carlbuchmann reviewed (approved):
LGTM
